trpcultivate_phenotypes/src/TripalCultivateValidator/ValidatorTraits/Headers.php

```diff
@@ -0,0 +1,220 @@
+  }
+
+  /**
+   * Get headers of a specified type(s).
+   * 
+   * @param array $types
+   *   A list of header types to get. 
+   *   Default to required and optional header types.
+   * 
+   * @return array
+   *   A single list of headers matching the type(s) defined by the 'types' parameter.
+   *   For example, if the 'types' parameter includes 'required' and 'optional', then the 
+   *   resulting array will contain all headers in order of index (this is assuming there
+   *   are only 'required' and 'optional' types supported).
+   *
+   *   The returned array is keyed by the index (column order) from the original headers 
+   *   array that was given to the setHeaders() method. 
+   *   The value is the header 'name'.
+   *   NOTE: the array is zero-based indexed.
+   * 
+   * @throws \Exception
+   * If the ['headers'][<one of the types given as a param>] key does not exist in the context array.
+   * This could be caused by either:
+   *  - The setHeaders() method has NOT been called.
+   *  - If an unrecognized header type is requested in the 'types' parameter.
+   */
+  public function getHeaders(array $types = ['required', 'optional']) {
+
+    $valid_types = $this->types;
+    
+    // This use of array_filter will pull out any unrecognized types by 
+    // comparing the 'types' parameter to the types property defined
+    // at the top of this class.
+    $invalid_types = array_filter($types, function($type) use($valid_types) { 
+      return !in_array($type, $valid_types); 
+    });
+    
+    // If any such unrecognized types are detected, throw an exception.
+    if (!empty($invalid_types)) {
+      $str_invalid_types = implode(', ', $invalid_types);
+      $str_valid_types = implode(', ', $valid_types);
+      throw new \Exception('Cannot retrieve invalid header types: ' . $str_invalid_types . '. Use one of valid types: [' . $str_valid_types . ']');
+    }
+   
+    if (array_key_exists($this->context_key, $this->context)) {
+      // At this point, types requested are valid.   
+      $headers = [];
+      $processed_types = 0;
+
+      foreach($types as $type) {
+        if (array_key_exists($type, $this->context[ $this->context_key ])) {
+          $processed_types++;
+          foreach($this->context[ $this->context_key ][ $type ] as $index => $header) {
+            $headers[ $index ] = $header;
+          }
+        }
+      }
+      
+      // All of the header elements with a type matching one of the types 
+      // requested, keyed by the column index and 'header name' as the value.
+      // Only return if all of the types requested were processed (i.e. were
+      // in the headers context array).
+      if ($processed_types == sizeof($types)) {
+        return $headers;
+      }
+    }
+    
+    throw new \Exception('Cannot retrieve headers from the context array as one has not been set by setHeaders() method.'); 
```

Comment:
I worry this message isn't specific enough to the range of problems that could have triggered it. Even though the most likely scenario is that setHeaders() wasn't called. But I don't want to suggest additional exceptions get thrown since they will be extremely tough to test 😆 How about this?
```suggestion
    throw new \Exception('Cannot retrieve requested headers as the ' . $this->context_key . ' key was not set in the context array, or one or more types were not set by setHeaders() method.'); 
```